### PR TITLE
Tim Roeder

### PR DESCRIPTION
**Pull request by tjroeder**
PR for Car Dealership independent challenge.

### CODE REVIEW

File: lib/car.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Car
+  attr_reader :make, :model, :monthly_payment, :loan_length, :total_cost, :color
+
+  def initialize(make_model, monthly_payment, loan_length)
+    @make = make_model.split[0]
+    @model = make_model.split[1]
+    @monthly_payment = monthly_payment
+    @loan_length = loan_length
+    @total_cost = monthly_payment * loan_length
```
Comment (BrianZanti) [inline lib/car.rb:9]: Rather than setting this as an attribute, I'd recommend making it a method that calculates this value every time we call it. What would happen if in the future we decided to make `make` and `model` attr_accessors?

File: lib/dealership.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  # Find the average price of cars in the inventory.
+  def average_price_of_car
+    # Assuming the average is under a million dollars.
+    average_price = (total_value / inventory_count)
+    if average_price >= 1000
+      average_price.to_s.insert(-4, ',')
+    else
+      average_price.to_s
+    end
```
Comment (BrianZanti) [inline lib/dealership.rb:48]: Nice job paying attention to details in the interaction pattern. We want this number to be formatted with commas. 

Something to consider is what if the total value was greater than 1,000,000? What if it was great than 1,000,000,000? Your logic would get more and more complex, and at some point it would probably be best to make a new Class or Module to handle that number formatting logic.

File: lib/dealership.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+  # Return sorted array of cars from cheapest to most expensive.
+  def cars_sorted_by_price
+    @inventory.sort_by(&:total_cost)
+  end
+
+  # Create a hash of inventory sorted by make.
+  def inventory_hash
+    @inventory.group_by(&:make)
```
Comment (BrianZanti) [inline lib/dealership.rb:58]: Excellent 🎉

File: lib/dealership.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    if average_price >= 1000
+      average_price.to_s.insert(-4, ',')
+    else
+      average_price.to_s
+    end
+  end
+
+  # Return sorted array of cars from cheapest to most expensive.
+  def cars_sorted_by_price
+    @inventory.sort_by(&:total_cost)
```
Comment (BrianZanti) [inline lib/dealership.rb:53]: I'd like to see you use the block syntax rather than this syntax just because we don't teach this syntax and it will be easier to communicate with your peers if you stay away from this.

File: spec/car_spec.rb [hunk 5]
```diff
@@ -2,39 +2,47 @@
...
+      expect(car.color).to eq(:blue)
+    end
+    
+    it 'can change the color again' do
+      car.paint!(:blue)
+      expect(car.color).to eq(:blue)
+      car.paint!(:green)
+      expect(car.color).to eq(:green)
+    end
   end
```
Comment (BrianZanti) [inline spec/car_spec.rb:47]: These tests look good! Another thing that would be good to test for is that the color starts as nil.

File: spec/dealership_spec.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      dealership.add_car(car_1)
+      expect(dealership.has_inventory?).to eq(false)
+      dealership.add_car(car_2)
+      expect(dealership.has_inventory?).to eq(true)
+      dealership.add_car(car_3)
+      expect(dealership.has_inventory?).to eq(true)
+      dealership.add_car(car_4)
+      expect(dealership.has_inventory?).to eq(true)
+    end
+  end
```
Comment (BrianZanti) [inline spec/dealership_spec.rb:70]: Great job testing all the different cases here!

File: spec/dealership_spec.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  
+  # Test Dealership#cars_by_make method.
+  describe '#cars_by_make' do
+    it 'can  return cars when given a make' do
+      dealership.add_car(car_1)
+      dealership.add_car(car_2)
+      dealership.add_car(car_3)
+      dealership.add_car(car_4)
+      expect(dealership.cars_by_make("Toyota")).to eq([car_2, car_3])
+      expect(dealership.cars_by_make("Ford")).to eq([car_1])
```
Comment (BrianZanti) [inline spec/dealership_spec.rb:80]: I see in a lot of your tests that you're doing a good job expecting the specific car objects that we want returned. Nice work 👍

File: spec/dealership_spec.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    end
+    
+    it 'can return the average price of the inventory' do
+      dealership.add_car(car_1)
+      dealership.add_car(car_2)
+      dealership.add_car(car_3)
+      dealership.add_car(car_4)
+      expect(dealership.average_price_of_car).to eq('39,000')
+    end
+  end
```
Comment (BrianZanti) [inline spec/dealership_spec.rb:120]: Again, nice job testing multiple cases.

File: spec/dealership_spec.rb [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  # Test Dealership#inventory_hash method.
+  describe '#inventory_hash' do
+    it 'can return a hash of the inventory sorted by make' do
+      dealership.add_car(car_1)
+      dealership.add_car(car_2)
+      dealership.add_car(car_3)
+      dealership.add_car(car_4)
+      expect(dealership.inventory_hash).to eq({'Ford'=> [car_1], 'Toyota'=> [car_2, car_3], 'Chevrolet'=> [car_4]})
+    end
+  end
```
Comment (BrianZanti) [inline spec/dealership_spec.rb:142]: Great work 👍

**Review by BrianZanti — COMMENTED** [review]
Hey Tim, excellent work on this challenge! You got everything we were looking for. I left you some things to think about but overall really great job 🎉

File: lib/dealership.rb (hunk 2 above)
Comment (tjroeder) [reply lib/dealership.rb:48]: Agreed. I did right a comment saying something along the lines of "assume average is less than a million." Mostly because it was outside of scope, and I really didn't want to write something to add a comma every thousands place. Initially I tried to find a method that could do something like that, insert every 3 places. But, all I came up with was to have an insert on a loop that has a counter that moves 3 places. I can probably come back to this in a refactor.

File: lib/dealership.rb (hunk 4 above)
Comment (tjroeder) [reply lib/dealership.rb:53]: Ok, no worries. I learn it recently and thought it was pretty slick.

File: spec/car_spec.rb (hunk 5 above)
Comment (tjroeder) [reply spec/car_spec.rb:47]: Dang, I totally missed that. Great catch. Thank you!
